core/unix/os.c

```diff
@@ -1087,6 +1087,15 @@ get_application_name_helper(bool ignore_cache, bool full_path)
     return (full_path ? executable_path : executable_basename);
 }
 
+#ifdef CLIENT_INTERFACE
+void
+set_client_error_code(dr_error_code_t error_code)
```

Comment:
I would expect this to be in instrument.c shared between OS's rather than duplicated in each OS impl file.

Reply:
Right, but I assume keeping the declaration in os_shared.h right? Otherwise, os.c cannot use it.

Reply:
Done.

Reply:
Actually, I am not sure how I can include set_client_error_code() in instrument.c and still use the function to set error codes in os.c and ntdll.c. Essentially, the function needs to have the same scope as get_thread_private_dcontext(). I tried moving the implementation to instrument.c and leaving the declaration in the os_shared.h but got a linking error on Appveyor.

I could not find a .c file that shares code between different OS's and is of lower dependency and instrument.c

Reply:
In the old days, a client_data_t operation like this would be declared in instrument.h and implemented in instrument.c.  os.c does include instrument.h.  If deeper parts of the core with no existing client interaction like ntdll.c needed to be used, a different layer would exist between ntdll.c and os.c which would use a different non-CLIENT_INTERFACE way to indicate errors, and the API function would convert that.

That said, these days nobody ever makes a non-CLIENT_INTERFACE build, we just live with whatever perf impact the define has, and we should probably clean up and simplify the code by removing the define altogether: #2971.

But does the Windows impl need to be in ntdll.c?  You could just put it in os.c and if os.c needs some ntdll data it can call into ntdll.c, and then have it in instrument.h and instrument.c  like the rest of the core client operations.

Reply:
Okay, understood. Thank you.